sdk/servicebus/service-bus/src/session/messageSession.ts

```diff
@@ -706,11 +725,24 @@ export class MessageSession extends LinkEntity<Receiver> {
             });
           }
         }
+
+        if (this.receiveMode === "peekLock" && numberOfEmptyIncomingSlots(this.link) <= 1) {
+          // Wait for the user to clear the deliveries before adding more credits
+          while (numberOfEmptyIncomingSlots(this.link) <= 1) {
```

Comment:
I'm not thrilled with this (having an arbitrary loop with a hardcoded delay). 

For instance, how does this loop get shut down? Does it go cleanly if the user closes the subscriber? etc...

An aside: I think with this change, given how infectious it is, I'd like to see us factor out this code (perhaps just into a common file) so it's obvious that all of these places are part of the same "feature". That at least can make it easier than seeing the changes intermixed without a clear tie between them.

Reply:
> I'm not thrilled with this (having an arbitrary loop with a hardcoded delay).

The hardcoded delay was just to show how it can be done, we can re-use existing delays or make a new const in the constants file for this delay.


> For instance, how does this loop get shut down? Does it go cleanly if the user closes the subscriber? etc...

I assumed the handler would be removed when the credits are drained. And the garbage collection would take care when the link is closed.
Very good question though.. 
- I need to test this case out - I'll add more tests and make sure they end gracefully.
- I had a `this.link` check in the while loop that I had removed during refactoring I guess, I can add the link.isOpen check back along with the canReceiveMessages check which would end the while loop gracefully.

Though the current solution is working, it's not fully optimized. 
I'm seriously loving these comments from you and @ramya-rao-a, I have great ideas to improve the current solution based on your comments. 💓

> I think with this change, given how infectious it is, I'd like to see us factor out this code (perhaps just into a common file) so it's obvious that all of these places are part of the same "feature"

Definitely, +1 for more refactoring.

Reply:
Will be done with https://github.com/Azure/azure-sdk-for-js/pull/14060#discussion_r620732468